# Vector Operations

I will be using the `cereal.csv` file, which was used in Chapter 3, as my data source. 


In [1]:
# import libraries
import pandas as pd
import numpy as np # this is used for numerical computing
import numpy.typing as npt # this is used for type annotation

In [2]:
# load data into dataframe
df = pd.read_csv('../datasets/cereal.csv')


Before I begin, I'd like to explain what vectors are.

Vectors are ordered list of numbers used to represent quantities or data mathematically. 

Vectors are the fundamental way machine learning represents and manipulates data. They are 1-dimensional structures.

Vectors are usually represented using NumPy arrays. (_Note that this does not mean all arrays are vectors_)

In this project, I will be representing various columns as vectors, and perform vector operations on them.

In [3]:
# represent the 'sugars' and 'calories' columns as  vectors
sugar_vector = np.array(df['sugars'])
calories_vector = np.array(df['calories'])

# check that they are both arrays
print(f"Sugar Vector is a {type(sugar_vector)}")
print(f"Calories Vector is a {type(calories_vector)}")

Sugar Vector is a <class 'numpy.ndarray'>
Calories Vector is a <class 'numpy.ndarray'>


The first question I'll be asking is: "Do cereals that have more sugar also tend to have more calories?"

To answer this question, I will be using Cosine Similarity

In [4]:
# create function to calculate cosine similarity
def cosine_similarity(A: npt.NDArray[np.float64], B: npt.NDArray[np.float64]) -> float:
    # compute dot product
    dot_prod = np.dot(A, B)

    # compute the norm of the vectors
    norm_A = np.linalg.norm(A)
    norm_B = np.linalg.norm(B)

    # find the product of the norms
    vector_norm = norm_A * norm_B

    # compute cosine similarity
    cos_sim = np.round(dot_prod / vector_norm, 3)

    return(cos_sim)



In [5]:
# print cosine similarity
cos_sim = cosine_similarity(sugar_vector, calories_vector)

print(f"The Cosine Similarity of the Sugar column and Calories column is {cos_sim}")

The Cosine Similarity of the Sugar column and Calories column is 0.883


_What does Cosine Similarity even mean?_

Cosine Similarity measures how similar two data items are by calculating the cosine of the angle between them. It focuses on the direction rather than magnitude, with a score of 1 meaning identical, 0 unrelated, and -1 opposite.


The Cosine Similarity of `0.883` means that there is a strong positive similarity between the two vectors.

This may suggest that cereals with higher sugar content generally have higher calorie values.


**NOTE:** 

Cosine Similarity does not measure correlation. Cosine similarity measures geometric similarity, while correlation measures statistical relationship.



**We have answered the question!**

Cereals that have more sugar tend to have more calories.

In [20]:
def vector_difference(cereal_1: str, cereal_2: str, nutrients: list[str]) -> dict[str, str]:
    """Calculates the difference between two vectors. """

    # Store original cereal names
    cereal_1_name = cereal_1
    cereal_2_name = cereal_2
    
    # Filter data for selected cereals, keep only numeric columns, then convert to NumPy float array
    cereal_1 = np.array(df[df["name"] == cereal_1_name]
                        .select_dtypes(include='number')).astype(float)
    cereal_2 = np.array(df[df["name"] == cereal_2_name]
                        .select_dtypes(include='number')).astype(float)
    
    # Create list containing nutrients
    nutrients = ['calories', 'protein', 'fat', 'sodium', 'fiber' ,'carbo', 'sugars', 'potass', 'vitamins']
    # Create list containing measurements
    measurement = ['calories', 'grams', 'grams', 'milligrams', 'grams' ,'grams', 'grams', 'milligrams', 'milligrams']
    # Calculate the difference in nutrients
    diff = cereal_1 - cereal_2
    
    # Loop through nutrients and check difference
    for n, d, m in zip(nutrients, diff[0], measurement):
        if d > 0:
            print(f"{cereal_1_name} has {abs(d)} {m} more {n} than {cereal_2_name}")
        elif d < 0:
            print(f"{cereal_1_name} has {abs(d)} {m} less {n} than {cereal_2_name}")
        else:
            print(f"{cereal_1_name} has the same amount of {n} as {cereal_2_name}")

vector_difference(cereal_1="All-Bran", cereal_2="Apple Jacks", nutrients=['protein', 'fat', 'sodium'])

All-Bran has 40.0 calories less calories than Apple Jacks
All-Bran has 2.0 grams more protein than Apple Jacks
All-Bran has 1.0 grams more fat than Apple Jacks
All-Bran has 135.0 milligrams more sodium than Apple Jacks
All-Bran has 8.0 grams more fiber than Apple Jacks
All-Bran has 4.0 grams less carbo than Apple Jacks
All-Bran has 9.0 grams less sugars than Apple Jacks
All-Bran has 290.0 milligrams more potass than Apple Jacks
All-Bran has the same amount of vitamins as Apple Jacks


In [8]:
# convert data frames to array
cereal_1 = np.array(df.iloc[5].drop(['name', 'mfr', 'type'])).astype(float)
cereal_2 = np.array(df.iloc[11].drop(['name', 'mfr', 'type'])).astype(float)

print(cereal_1 - cereal_2)

df.columns

[   0.         -4.          0.       -110.         -0.5        -6.5
    9.        -35.          0.          0.          0.         -0.5
  -21.255458]


Index(['name', 'mfr', 'type', 'calories', 'protein', 'fat', 'sodium', 'fiber',
       'carbo', 'sugars', 'potass', 'vitamins', 'shelf', 'weight', 'cups',
       'rating'],
      dtype='object')